src/Catalyst.Core.Modules.Dfs/BlockExchange/BitswapService.cs

```diff
@@ -0,0 +1,545 @@
+
+        /// <summary>
+        ///   Indicate that a remote peer sent a block.
+        /// </summary>
+        /// <param name="remote">
+        ///   The peer that sent the block.
+        /// </param>
+        /// <param name="block">
+        ///   The data for the block.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///   <para>
+        ///   Updates the statistics.
+        ///   </para>
+        ///   <para>
+        ///   If the block is acceptable then the <paramref name="block"/> is added to local cache
+        ///   via the <see cref="BlockService"/>.
+        ///   </para>
+        /// </remarks>
+        public Task OnBlockReceivedAsync(Peer remote, byte[] block)
+        {
+            return OnBlockReceivedAsync(remote, block, Cid.DefaultContentType, MultiHash.DefaultAlgorithmName);
+        }
+
+        /// <summary>
+        ///   Indicate that a remote peer sent a block.
+        /// </summary>
+        /// <param name="remote">
+        ///   The peer that sent the block.
+        /// </param>
+        /// <param name="block">
+        ///   The data for the block.
+        /// </param>
+        /// <param name="contentType">
+        ///   The <see cref="Cid.ContentType"/> of the block.
+        /// </param>
+        /// <param name="multiHash">
+        ///   The multihash algorithm name of the block.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///   <para>
+        ///   Updates the statistics.
+        ///   </para>
+        ///   <para>
+        ///   If the block is acceptable then the <paramref name="block"/> is added to local cache
+        ///   via the <see cref="BlockService"/>.
+        ///   </para>
+        /// </remarks>
+        public async Task OnBlockReceivedAsync(Peer remote, byte[] block, string contentType, string multiHash)
+        {
+            // Update statistics.
+            ++BlocksReceived;
+            DataReceived += (ulong) block.LongLength;
+            peerLedgers.AddOrUpdate(remote,
+                (peer) => new BitswapLedger
+                {
+                    Peer = peer,
+                    BlocksExchanged = 1,
+                    DataReceived = (ulong) block.LongLength
+                },
+                (peer, ledger) =>
+                {
+                    ++ledger.BlocksExchanged;
+                    DataReceived += (ulong) block.LongLength;
+                    return ledger;
+                });
+
+            // TODO: Detect if duplicate and update stats
+            var isDuplicate = false;
+            if (isDuplicate)
+            {
+                ++DupBlksReceived;
+                DupDataReceived += (ulong) block.Length;
+            }
+
+            // TODO: Determine if we should accept the block from the remote.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)